### PR TITLE
Improvements with the localizer.

### PR DESCRIPTION
**Pull request by thibaultzanini**
Improvements with the Localizer.
- Resolves an issue that was not update the messages when the locale property of the defaultLocalizer changes.
- Add a method `localizerDidChangeLocale` that can be implemented by components in order to be aware when the locale property of the defaultLocalizer changes.
- Add a delegate method `localizerWillLoadMessages` that can provide a ”messages" object with a combination of keys/messages.
- Add a static method `defaultLocalizerWithDelegate` on Localizer in order to set the delegate of the defaultLocalizer. (Returns the defaultLocalizer)
- Add a static method `defaultLocalizer` on Localizer in order to get the defaultLocalizer.
- Merge “class” Localizer and DefaultLocalizer.
- Add a flag to activate the locale storage, `shouldStoreLocale`.
- Tests updated.

### CODE REVIEW

File: core/localizer.js [hunk 1]
```diff
@@ -250,19 +282,30 @@ var Localizer = exports.Localizer = Montage.specialize( /** @lends Localizer.pro
...
-            var promise = this._manifest;
+            var delegate = this.delegate,
+                self = this,
+                promise;
 
-            if (timeout) {
-                promise = promise.timeout(timeout);
-            }
+            if (delegate && typeof delegate.getLocaleMessages === "function") {
+                promise = delegate.getLocaleMessages(this.locale).timeout(timeout);
```
Comment (thibaultzanini) [inline core/localizer.js]: localizerWillLoadMessages

File: core/localizer.js [hunk 2]
```diff
@@ -250,19 +282,30 @@ var Localizer = exports.Localizer = Montage.specialize( /** @lends Localizer.pro
...
-            var self = this;
-            var promise = this._manifest;
+            var delegate = this.delegate,
+                self = this,
+                promise;
 
-            if (timeout) {
-                promise = promise.timeout(timeout);
-            }
+            if (delegate && typeof delegate.getLocaleMessages === "function") {
```
Comment (thibaultzanini) [inline core/localizer.js]: use callDelegateMethod

File: core/localizer.js [hunk 3]
```diff
@@ -598,6 +627,34 @@ var DefaultLocalizer = Localizer.specialize( /** @lends DefaultLocalizer# */ {
         }
     },
 
+    _dispatchLocaleChange: {
+        value: function (previousLocale, component) {
```
Comment (thibaultzanini) [inline core/localizer.js]: _component

File: core/localizer.js [hunk 4]
```diff
@@ -598,6 +627,34 @@ var DefaultLocalizer = Localizer.specialize( /** @lends DefaultLocalizer# */ {
...
+            if (!component) {
+                component = rootComponent; //@todo: ask benoit if it's needed for root component ?
```
Comment (thibaultzanini) [inline core/localizer.js]: add a comment

**Comment by thibaultzanini** [comment]
@marchant PR updated

**Comment by thibaultzanini** [comment]
@marchant PR updated

File: core/localizer.js [hunk 5]
```diff
@@ -497,120 +574,124 @@ var Localizer = exports.Localizer = Montage.specialize( /** @lends Localizer.pro
...
-                    window.localStorage.setItem(LOCALE_STORAGE_KEY, value);
-                } catch (e) {
-                    // LocalStorage quota might have been exceeded
-                    // iOS Safari emits a quota exceeded error from private mode always
-                }
+    defaultLocalizer: {
+        value: function () {
+            if (!defaultLocalizer) {
+                defaultLocalizer = new Localizer();
+                defaultLocalizer.shouldStoreLocale = true;
```
Comment (marchant) [inline core/localizer.js]: shouldStoreLocale should be storesLocale, shouldStoreLocale would be a proper name for a delegate method

File: core/localizer.js [hunk 6]
```diff
@@ -497,120 +574,124 @@ var Localizer = exports.Localizer = Montage.specialize( /** @lends Localizer.pro
...
-     */
-    reset: {
-        value: function () {
-            if (typeof window !== "undefined" && window.localStorage) {
-                window.localStorage.removeItem(LOCALE_STORAGE_KEY);
+    defaultLocalizerWithDelegate: {
+        value: function (delegate) {
+            if (!defaultLocalizer) {
+                defaultLocalizer = new Localizer();
+                defaultLocalizer.shouldStoreLocale = true;
```
Comment (marchant) [inline core/localizer.js]: This should call this.defaultLocaliser()

File: core/localizer.js [hunk 7]
```diff
@@ -497,120 +574,124 @@ var Localizer = exports.Localizer = Montage.specialize( /** @lends Localizer.pro
...
-                if (window.localStorage) {
-                    defaultLocale = window.localStorage.getItem(LOCALE_STORAGE_KEY);
+
+    _dispatchLocaleChange: {
+        value: function (previousLocale, _component) {
+            if (!_component) {
+                _component = this.component || rootComponent;
+
+                if (typeof _component.LocalizerDidChangeLocale === "function") {
+                    _component.LocalizerDidChangeLocale(this, previousLocale, this._locale);
```
Comment (marchant) [inline core/localizer.js]: this should be localizerDidChangeLocale, lowercase "l". Symbol names starting by a capital letter are usully reserved for Types/classes, constants

File: core/localizer.js [hunk 8]
```diff
@@ -497,120 +574,124 @@ var Localizer = exports.Localizer = Montage.specialize( /** @lends Localizer.pro
...
-        }
-    },
+                for (var i = 0; i < childComponents.length; i++) {
+                    child = childComponents[i];
 
-    _delegate: {
-        value: null
-    },
+                    if (typeof child.LocalizerDidChangeLocale === "function") {
+                        child.LocalizerDidChangeLocale(this, previousLocale, this._locale);
```
Comment (marchant) [inline core/localizer.js]: localizerDidChangeLocale

File: core/localizer.js [hunk 9]
```diff
@@ -47,38 +48,90 @@ var reLanguageTagValidator = /^[a-zA-Z]+(?:-[a-zA-Z0-9]+)*$/;
...
+     * @param {string} [locale] The RFC-5646 language tag this localizer should use. Defaults to defaultLocalizer.locale
      * @returns {Localizer} The Localizer object it was called on.
      */
-    init: {
+    initWithLocale: {
         value: function (locale) {
-            this.locale = locale || defaultLocalizer.locale;
+            var defaultLocale = this.callDelegateMethod("localizerWillUseLocale");
+
+            if (!defaultLocale && this.shouldStoreLocale && typeof window !== "undefined" && window.localStorage) {
```
Comment (marchant) [inline core/localizer.js]: should be this.storesLocale

File: core/localizer.js [hunk 10]
```diff
@@ -47,38 +48,90 @@ var reLanguageTagValidator = /^[a-zA-Z]+(?:-[a-zA-Z0-9]+)*$/;
...
+    initWithLocaleAndDelegate: {
+        value: function (locale, delegate) {
+            this.delegate = delegate;
+
+            return this.initWithLocale(locale);
+        }
+    },
+
+
+    shouldStoreLocale: {
```
Comment (marchant) [inline core/localizer.js]: storesLocale

File: core/localizer.js [hunk 11]
```diff
@@ -47,38 +48,90 @@ var reLanguageTagValidator = /^[a-zA-Z]+(?:-[a-zA-Z0-9]+)*$/;
...
-     * @param {string} [locale] The RFC-5646 language tag this localizer
-     * should use. Defaults to defaultLocalizer.locale
+     * @param {string} [locale] The RFC-5646 language tag this localizer should use. Defaults to defaultLocalizer.locale
      * @returns {Localizer} The Localizer object it was called on.
      */
-    init: {
+    initWithLocale: {
         value: function (locale) {
-            this.locale = locale || defaultLocalizer.locale;
+            var defaultLocale = this.callDelegateMethod("localizerWillUseLocale");
```
Comment (marchant) [inline core/localizer.js]: How is locale passed to the delegate here?

File: core/localizer.js (hunk 8 above)
Comment (marchant) [inline core/localizer.js]: Spec showing the use of aComponent.LocalizerDidChangeLocale?

**Comment by marchant** [comment]
Reviewed, LGTM

**Pull request merged**